### DISCUSSION TITLE
How to get the new node to show up in `useNodesState` for modification?

### DISCUSSION THREAD

**Discussion by CSworenI** [xyflow/xyflow#4512] [Q&A] [answered]
I'm using a version of the [drag and drop example](https://reactflow.dev/examples/interaction/drag-and-drop) and I'm not able to see the new nodes when I console log `edges`. Also, modifying an edge connecting with it is failing because I'm getting "Cannot read properties of undefined" errors related to the new node that was just dropped in. What am I doing wrong? It's a critical piece of functionality for building a flow chart that seems to be <ins>**completely broken**</ins> (I even tried console logging and modifying the examples which create new nodes and everything fails).

Here's what I'm doing:
```
const { addNodes } = useReactFlow(),

onDragOver = useCallback(event => {
	event.preventDefault();
	event.dataTransfer.dropEffect = 'move';
}, []),

onDrop = useCallback(event => {
	event.preventDefault();
	const type = event.dataTransfer.getData('application/reactflow');
	if (typeof type === 'undefined' || !type) { return; }
	const position = flowInstance.screenToFlowPosition({ x: event.clientX, y: event.clientY }),
		newNode = { id: getId(), type, position };
	switch (type) {
		case 'end':
			newNode.type = 'process';
			newNode.data = { status: 'START', label: 'new start/end node' };
			newNode.style = ENDSTYLE;
			break;
		case 'process':
			newNode.type = type;
			newNode.data = { status: 'PLANNED', label: 'new process (tasks) node' };
			newNode.bg = '#dbeafe';
			break;
		case 'decision':
			newNode.type = type;
			newNode.data = { label: 'new decision node' };
			newNode.bg = '#ffa3a6';
			break;
	}
	addNodes(newNode);
	console.log(nodes); // always the same list of initial nodes, even after 10+ nodes dropped in
}, [flowInstance]),
```

I also tried `setNodes(n => n.concat(newNode));` as recommended practically everywhere, without success.

**Comment by bcakmakoglu** [comment]


**Reply by CSworenI** [reply]
@bcakmakoglu This doesn't work either. I'm still getting the "Cannot read properties of undefined" error when attempting to modify the edge connecting to a node that wasn't in my initial nodes list. Same when attempting to open a modal on click to display the attributes of the new node. It doesn't exist in the `nodes` list.

I'm trying to achieve a way of adding into the `nodes` list that is propagated with `const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes)`. I don't have this issue with `[edges, setEdges, onEdgesChange] = useEdgesState(initialEdges)` and persisting changes to it, thanks to `{ updateEdge } = useReactFlow()`. Does a similar construct exist for nodes?

It looks like the workaround is to have  `const [myNodes, setMyNodes] = useState(initialNodes)` to hold changes and hijack the `setMyNodes` to also feed into `setNodes` so the changes propagate to the `ReactFlow` object. This might be OK with smaller flow charts, but most of the ones anticipated will be hundreds of nodes, which will wind up slowing down at some point, maybe drastically.

**Reply by CSworenI** [reply]
Also, your comment about the library being mean to me is rude and unprofessional. My day-to-day has me working regularly with more than a dozen technologies, including other frontend libraries, so subtle nuances and deep inner workings are difficult to keep straight everywhere. I openly admitted `I don't know what I'm doing wrong` and stated it 'seems' to be broken functionality.

**Reply by CSworenI** [reply]
Figured it out without having all that extra crap...I needed to manipulate the original array of `initialNodes`, then use `setNodes` to ensure it's visible in `nodes`. It's not as clean a solution as I'd prefer, but it works. Let me know if you know a shorter way to make new nodes and node updates visible. I'll mark this as answer if you don't come up with something by Tuesday EOD.

**Reply by bcakmakoglu** [reply]
It wasn't meant to be rude. Though after re-reading I can see how it came off that way, apologies - I removed the comment since you seem to have found your own solution.

Have a nice day.

**Reply by CSworenI** [reply, marked as answer]
That workaround wasn't capturing user interaction changes. The real answer is using `setNodes([ ...nodes, newNode ]);` to add to the array and using a dependency on `nodes` to capture the change.
https://github.com/xyflow/xyflow/discussions/4242#discussioncomment-10243351
